### ISSUE TITLE
Compiler warnings with gcc-9 and --enable-ztex

### ISSUE THREAD

**Issue by AlbertVeli** [openwall/john#4220]
### Steps to reproduce ###
./configure --enable-ztex

### Warnings ###
See:
https://gist.github.com/AlbertVeli/8727d0924a594a1ff50a2d4e1e56c1eb

### Notes ###
The warnings come from the format string "%s_%d" where the %s part can be up to 255 bytes and the destination has 256 bytes allocated. To actually perform an overflow _device->ztex_device->snString_ has to be manipulated in some way. Maybe possible by editing the config file (or even more far fetched, attaching a malicious ztex-board that reports a 255 character serial number).

Far fast reproduce, just:
```
cd ztex
gcc -c -Wall -O2 device.c
```
I have only tried with gcc-9.2.

**Comment by ghost** [comment]
JtR issues many alerts on new compilers [1]. E.g.,
* ztex on 9.0
https://travis-ci.org/claudioandre-br/JohnTheRipper/jobs/657887727

If possible, Solar should think about it (and how/if we can fix it).

[1] gcc and clang, All recent versions.

**Comment by magnumripper** [comment]
These seem to be compiler bugs. Is this a stable gcc version?

Example from luks_fmt:
```c
static int crypt_all(int *pcount, struct db_salt *salt)
{
	const int count = *pcount;
	int index;

#ifdef _OPENMP
#pragma omp parallel for
#endif
	for (index = 0; index < count; index += MIN_KEYS_PER_CRYPT) {
		(...)
		int i, iterations = cur_salt->bestiter;
		(...)

		for (i = 0; i < MIN_KEYS_PER_CRYPT; ++i) {
			lens[i] = strlen(saved_key[index+i]);
```
```
luks_fmt_plug.c:557:31: warning: variable 'index' is uninitialized when used here [-Wuninitialized]
                        lens[i] = strlen(saved_key[index+i]);
                                                   ^~~~~
luks_fmt_plug.c:537:11: note: initialize the variable 'index' to silence this warning
```
Now tell me, how the heck is index not initialized? Looks to me the compiler made buggy optimizations and then complains about the result 🙄 


And anyway, perhaps it's time we drop `-Wall` for a standard build? The compiler vendors have been **begging** us for it for years.

**Comment by magnumripper** [comment]
```
clang version: 9.0.0 (tags/RELEASE_900/final)
```
Looks stable all right. Could still be a compiler bug of course. I can't fathom what else it could be.

**Comment by magnumripper** [comment]
I just tried a ZTEX build using Ubuntu 18.04 (which has gcc 7.4.0). Not a single warning, ASan or not.

**Comment by magnumripper** [comment]
Just for the hell of it I tried actually NOT initializing `index` and then I got the (now correct) warning. I had to drop the `#pragma omp` for gcc to accept it at all.

**Comment by magnumripper** [comment]
Same Ubuntu 18.04 with gcc-8, no problem. clang-9, problem confirmed.

**Comment by magnumripper** [comment]
Seems to be a bug with their OpenMP stuff: `--disable-openmp` gets rid of every single warning.

**Comment by magnumripper** [comment]
Perhaps using the tedious full syntax would fix it.
```c
#pragma omp parallel for default(none) private(index) shared(count, saved_key, saved_len, (...))
```

If we're lucky we might get away with
```c
#pragma omp parallel for default(shared) private(index)
```
EDIT: That's not likely. And using any `default` should be far more bug-prone than our current way of not declaring at all. I've never seen any gcc bug with this.

**Comment by magnumripper** [comment]
I believe the OP is an actual issue though.

**Comment by AlbertVeli** [comment]
> I believe the OP is an actual issue though.

This issue is not severe but it would feel good to get rid of the warning.

Do you run john through Coverity scans? I think free/open source projects on github can get free Coverity scans and it should find most of the things gcc/clang can find, and more issues too that future gcc versions might warn about, through their static code analysis.

**Comment by magnumripper** [comment]
> Do you run john through Coverity scans?

We've done so in the past, the false-to-real ratio use to be like 25:1 so it's not for everyday use.

**Comment by solardiz** [comment]
Albert, maybe you can be the one to patch the ZTEX code warnings issue you reported here? Please send us a PR.

Do you have some of those boards now?

On Coverity:

We also made some use of Coverity on smaller and cleaner Openwall projects - blists and passwdqc - and even there the false positives were a problem... and there were false negatives too (Coverity missed use of a potentially uninitialized variable in older passwdqc, presumably because this was over translation unit boundary). I also see how the libxcrypt project uses Coverity and then tries to modify/complicate the code to avoid the false positives - that's quite counter-productive.

I don't see Coverity helping JtR jumbo given the present shape of both Coverity and JtR jumbo.

**Comment by AlbertVeli** [comment]
Sure. I can try to fix the compiler warnings. Someone should audit before merging later because I am not very familiar with the code.

And yes, I bought three ZTEX boards from the ebay link you posted: https://www.openwall.com/lists/john-users/2020/02/12/3 The cards works well, I just had to replace some fans (and remove some dust). I also got spacers and cables with the boards.

**Comment by solardiz** [comment]
@AlbertVeli Sounds great. Maybe you can post to that john-users thread sharing your experience buying from that seller and testing with JtR. Thanks!

**Comment by solardiz** [comment]
The original issue reported here is fixed with #4222.

magnum, you might want to open another issue for the unrelated gcc 9 warnings you were talking about here.
